File: winui/CellposeGUI/Models/SessionState.cs
using CellposeGUI.Models;

namespace CellposeGUI.Services;

public sealed class SessionState
{
    public string? ImagePath { get; set; }
    public ImageData? Image { get; set; }
    public MaskData? Masks { get; set; }
    public List<ArrayPayload> Flows { get; set; } = [];
    public bool RecomputeMasks { get; set; }
    public string Model { get; set; } = "cpsam";
    public SegmentationParameters Segmentation { get; set; } = new();
    public bool[] IsManual { get; set; } = [];
    public SeriesDatasetPayload? SeriesDataset { get; set; }

    public int Ncells => Masks?.Labels.Length > 0 ? Masks.Labels.Max() : 0;

    public bool HasImage => Image != null && ImagePath != null;
    public bool HasMasks => Masks != null && Ncells > 0;
}

File: swiftui/CellposeGUI/Models/SessionState.swift
import CellposeGUICore
import Foundation

final class SessionState {
    var imagePath: String?
    var image: ImageData?
    var masks: MaskData?
    var flows: [ArrayPayload] = []
    var recomputeMasks = false
    var model = "cpsam"
    var segmentation = SegmentationParameters()
    var seriesDataset: SeriesDatasetPayload?

    var ncells: Int {
        guard let labels = masks?.labels, !labels.isEmpty else { return 0 }
        return Int(labels.max() ?? 0)
    }

    var hasImage: Bool { image != nil && imagePath != nil }
    var hasMasks: Bool { ncells > 0 }
}

struct LoadedSession {
    let imagePath: String
    let image: ImageData
    let masks: MaskData
    let flows: [ArrayPayload]
    let recomputeMasks: Bool
    let model: String
    let segmentation: SegmentationParameters
}

struct InferResult {
    let masks: MaskData?
    let flows: [ArrayPayload]
    let ncells: Int
    let recomputeMasks: Bool
}

struct RecomputeResult {
    let masks: MaskData?
    let ncells: Int
}
